roaring/roaring.go

```diff
@@ -2289,7 +2292,7 @@ func (c *container) bitmapSetRange(i, j uint64) {
 	x := i >> 6
 	y := (j - 1) >> 6
 	var X uint64 = maxBitmap << (i % 64)
-	var Y uint64 = maxBitmap >> (64 - (j % 64))
+	var Y uint64 = maxBitmap >> (63 - ((j - 1) % 64))
```

Comment:
can you look at algorithm 3 on page 11 of the roaring paper: https://arxiv.org/pdf/1603.06549.pdf

Is there an error in the paper's algorithm, or is there some subtle difference between this implementation values/variables in the paper?

Reply:
The problem that I see is when `j` is a multiple of 64, then instead of shifting `maxBitmap` by 0 it shifts it by 64. Notice the ranges for `Y` (Paper vs. Adjusted). I don't understand the case where the algorithm in the paper works on the edge.
```
i:                [0-63]
j:                [1-64]
x = i >> 6:       [0-63]
y = (j - 1) >> 6: [0-63]

X = max << (i % 64)
  = max << ([0-63] % 64)
  = max << [0-63]

Paper:
Y = max >> (64 - (j % 64))
  = max >> (64 - ([1-64] % 64))
  = max >> (64 - [1-63,0])
  = max >> [63-1,64])

Adjusted:
Y = max >> (63 - ((j - 1) % 64))
  = max >> (63 - (([1-64] - 1) % 64))
  = max >> (63 - (([0-63]) % 64))
  = max >> (63 - ([0-63])
  = max >> [63-0]
```

Reply:
Yeah, I guess the paper is wrong - I just did a direct translation of the paper algorithm to Go and reproduced the problem (setting i=0 and j=64 doesn't set any bits).

I also noticed a few other discrepancies:
- The paper defines "bitmap B, as an array of 64-bit words C_1 to C_1024", but line 11 references C_0 if i < 64. So I think it should be defined as C_0 to C_1023.
- The `for` loop on line 12 uses `i` as a loop variable which is confusing since `i` is also an argument.
- Inside the loop on line 13 C_x is used, but it should be C_i (the loop variable) otherwise it would operate on C_x every time.

Paging @lemire

Reply:
Here is my direct port of Algorithm 3 to go, with a demonstration of the issue: https://play.golang.org/p/nVBMYAtsyCK

Reply:
yeah, and here is the "official go port"
```

Reply:
well that's part of it - I found this func 
```go
func setBitmapRange(bitmap []uint64, start int, end int) {
	if start >= end {
		return
	}
	firstword := start / 64
	endword := (end - 1) / 64
	if firstword == endword {
		bitmap[firstword] |= (^uint64(0) << uint(start%64)) & (^uint64(0) >> (uint(-end) % 64))
		return
	}
	bitmap[firstword] |= ^uint64(0) << uint(start%64)
	for i := firstword + 1; i < endword; i++ {
		bitmap[i] = ^uint64(0)
	}
	bitmap[endword] |= ^uint64(0) >> (uint(-end) % 64)
}
``` 
which does behave properly. that `uint(-end)` bit is a neat trick.